File: binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
           
        left = get_height(root.left)
        right = get_height(root.right)
        if abs(left - right) > 1:
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)

def get_height(node):
    if node is None:
        return 0
    return 1 + max(get_height(node.left), get_height(node.right))

File: binary-tree/balanced-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end

def get_max_height(node)
    return -Float::INFINITY if node.nil?
    is_leaf = node.left.nil? && node.right.nil?
    return 1 if is_leaf
    left_height = get_max_height(node.left)
    right_height = get_max_height(node.right)
    return 1 + [left_height, right_height].max
end

def get_min_height(node)
    return Float::INFINITY if node.nil?
    is_leaf = node.left.nil? && node.right.nil?
    return 1 if is_leaf
    left_height = get_min_height(node.left)
    right_height = get_min_height(node.right)
    return 1 + [left_height, right_height].min
end

# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
    return true if root.nil?
    (get_max_height(root) - get_min_height(root)).abs <= 1
end